core/system/src/pool_queue.c

```diff
@@ -419,3 +420,170 @@ uvisor_pool_slot_t uvisor_pool_queue_find_first(uvisor_pool_queue_t * pool_queue
 
     return slot;
 }
+
+static bool pool_sane(uvisor_pool_t * pool, void * array,
+                      size_t stride, size_t num)
+{
+    /* Invalid num can never be sane */
+    if (num == 0 || num >= UVISOR_POOL_MAX_VALID) {
+        return false;
+    }
+
+    /* Check magic value */
+    if (pool->magic != UVISOR_POOL_MAGIC) {
+        return false;
+    }
+
+    /* Check pool size */
+    if (pool->num != num) {
+        return false;
+    }
+
+    /* Check array pointer */
+    if (pool->array != array) {
+        return false;
+    }
+
+    /* Check stride */
+    if (pool->stride != stride) {
+        return false;
+    }
+
+    /* Allocated count cannot be larger than size */
+    if (pool->num_allocated > num) {
+        return false;
+    }
+
+    /* Use a larger uint size to prevent overflow wraparound */
+    uint32_t num_free = 0;
+    uvisor_pool_slot_t i = pool->first_free;
+
+    /* Cycle detection: If there has been a cycle in the free list and there is
+     * an upper bound for num_free and a check of allocated count can indicate
+     * a cycle or corrupted value */
+    while (i < num && num_free < num) {
+        uvisor_pool_queue_entry_t * slot = &pool->management_array[i];
+        /* Check state */
+        if (slot->dequeued.state != UVISOR_POOL_SLOT_IS_FREE) {
+            return false;
+        }
+        num_free++;
+        i = slot->dequeued.next;
+    }
+
+    /* Check the whether the last element in the free list has correct `next`
+     * value */
+    if (i != UVISOR_POOL_SLOT_INVALID) {
+        return false;
+    }
+
+    /* Check allocated count */
+    if (pool->num_allocated + num_free != num) {
+        return false;
+    }
+
+    return true;
+}
+
+int uvisor_pool_sanity_check(uvisor_pool_t * pool, void * array,
+                         size_t stride, size_t num)
+{
+    if (!uvisor_spin_trylock(&pool->spinlock)) {
+        return UVISOR_ERROR_CANNOT_LOCK;
+    }
+    bool is_sane = pool_sane(pool, array, stride, num);
+    uvisor_spin_unlock(&pool->spinlock);
+    return is_sane ? 0 : UVISOR_ERROR_OUT_OF_STRUCTURES;
+}
+
+int uvisor_pool_queue_sanity_check(uvisor_pool_queue_t * queue,
+                               uvisor_pool_t * pool, void * array,
+                               size_t stride, size_t num)
+{
+    bool visited[num];
+    int status = UVISOR_ERROR_OUT_OF_STRUCTURES;
+    /* Cannot check sanity if lock is held by others */
+    if (!uvisor_spin_trylock(&queue->pool->spinlock)) {
```

Comment:
It seems we are locking too early, over too much. Please lock minimally, just around the accesses or cluster of accesses where locking is needed.

Reply:
Locking will not be effective, as a malicious user can always modify the internals ignore the lock

Reply:
The locking is to prevent us from seeing the pool or queue inconsistently and mistakenly judging it to be defective.